lib/grape/middleware/base.rb

```diff
@@ -27,7 +31,20 @@ def call!(env)
         ensure
           after_response = after
         end
-        after_response || @app_response
+
+        response = after_response || @app_response
+        reflect_headers response
+        response
+      end
+
+      def reflect_headers(response)
+        return unless headers.is_a?(Hash)
+        response_headers =
```

Comment:
I am nitpicking but I would avoid a temporary `response_headers` and just do `when .... .merge!(...)`.